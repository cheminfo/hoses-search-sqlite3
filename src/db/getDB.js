import fs, { existsSync, mkdirSync } from 'fs';
import { join } from 'path';

import sqLite from 'better-sqlite3';

let db;

export default function getDB() {
  const path = new URL('../../sqlite/', import.meta.url).pathname;
  console.log({ path })
  if (!db) {
    if (!existsSync(path)) {
      mkdirSync(path);
    }
    db = sqLite(join(path, 'db.sqlite'));
    // https://www.sqlite.org/wal.html
    // Activating WAL mode allows to get a speed improvement of 100x !!!
    db.pragma('journal_mode = WAL');
  }

  setInterval(() => {
    fs.stat(join(path, 'db.sqlite-wal'), (err, stat) => {
      if (err) {
        if (err.code !== 'ENOENT') throw err;
      } else if (stat.size > 100000000) {
        db.pragma('wal_checkpoint(RESTART)');
      }
    });
  }, 300000).unref();

  const createTable = `
CREATE TABLE IF NOT EXISTS molecules (
  id data_type PRIMARY KEY,
  idCode data_type TEXT,
  ssIndex0 data_type INTEGER,
  ssIndex1 data_type INTEGER,
  ssIndex2 data_type INTEGER,
  ssIndex3 data_type INTEGER,
  ssIndex4 data_type INTEGER,
  ssIndex5 data_type INTEGER,
  ssIndex6 data_type INTEGER,
  ssIndex7 data_type INTEGER
);
CREATE INDEX IF NOT EXISTS ssIndex0 ON molecules(ssIndex0);
CREATE INDEX IF NOT EXISTS ssIndex1 ON molecules(ssIndex1);
CREATE INDEX IF NOT EXISTS ssIndex2 ON molecules(ssIndex2);
CREATE INDEX IF NOT EXISTS ssIndex3 ON molecules(ssIndex3);
CREATE INDEX IF NOT EXISTS ssIndex4 ON molecules(ssIndex4);
CREATE INDEX IF NOT EXISTS ssIndex5 ON molecules(ssIndex5);
CREATE INDEX IF NOT EXISTS ssIndex6 ON molecules(ssIndex6);
CREATE INDEX IF NOT EXISTS ssIndex7 ON molecules(ssIndex7);
`;

  db.exec(createTable);
  return db;
}
