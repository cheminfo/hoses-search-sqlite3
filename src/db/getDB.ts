import { existsSync, mkdirSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

import { v4 } from '@lukeed/uuid';
import sqLite, { Database } from 'better-sqlite3';
import Postgrator from 'postgrator';

const __dirname = dirname(fileURLToPath(import.meta.url));

let db: Database;

export async function getDB(): Promise<Database> {
  if (!db?.open) {
    const path = new URL('../../sqlite/', import.meta.url).pathname;
    if (!existsSync(path)) {
      mkdirSync(path);
    }
    db = sqLite(join(path, 'db.sqlite'));
    // https://www.sqlite.org/wal.html
    // Activating WAL mode allows to get a speed improvement of 100x !!!
    db.pragma('journal_mode = WAL');
  }
  await prepareDB(db);
  return db;
}

export function getAllTableInfo(db: Database) {
  const sql = `
SELECT name, sql
FROM sqlite_master
WHERE type='table';`;
  const stmt = db.prepare(sql);
  const rows = stmt.all();
  return rows;
}

export async function getTempDB(): Promise<Database> {
  const tempDB = sqLite(`file:${v4()}?mode=memory`);
  await prepareDB(tempDB);
  return tempDB;
}

export async function prepareDB(db: Database) {
  const postgrator = new Postgrator({
    migrationPattern: join(__dirname, 'migrations/*'),
    driver: 'sqlite3',
    execQuery: (query) => {
      return new Promise((resolve, reject) => {
        const stmt = db.prepare(query);
        try {
          const rows = stmt.all();
          resolve({ rows });
        } catch (error) {
          //@ts-expect-error This is a hack to avoid the error
          if (error.message.includes('This statement does not return data')) {
            stmt.run();
            resolve({ rows: [] });
          }
          throw error;
        }
      });
    },
    execSqlScript: (sqlScript) => {
      return new Promise((resolve, reject) => {
        db.exec(sqlScript);
        resolve();
      });
    },
  });
  await postgrator.migrate();
}
