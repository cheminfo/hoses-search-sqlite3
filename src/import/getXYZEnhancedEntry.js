import OCL from 'openchemlib';
import { TopicMolecule } from 'openchemlib-utils';

import { convertXYZToMolfile } from '../qm9/utils/convertXYZToMolfile.js';

const { Molecule } = OCL;

export async function getXYZEnhancedEntry(lines, options = {}) {
  const { xyz, contact } = options;
  const entry = {};
  entry.nbAtoms = Number.parseInt(lines[0]);
  entry.comment = lines[1];
  entry.atoms = [];
  const cleanLines = lines.slice(0);
  const atoms = [];
  for (let i = 2; i < entry.nbAtoms + 2; i++) {
    const line = lines[i];
    const [atomSymbol, x, y, z, ...properties] = line.split(/\s+/);
    const atom = { atomSymbol, x, y, z, properties: properties.map(Number) };
    if (xyz?.columns) {
      for (let i = 0; i < atom.properties.length; i++) {
        atom.properties[i] = {
          energy: atom.properties[i],
          algorithm: xyz.columns[i].algorithm,
          orbital: xyz.columns[i].orbital,
          contact: contact.email,
        };
      }
    }
    atoms.push(atom);
    cleanLines[i] = `${atomSymbol} ${x} ${y} ${z}`;
  }
  entry.xyz = cleanLines.join('\n');
  entry.molfile3D = await convertXYZToMolfile(entry.xyz);
  const { molecule, map } = Molecule.fromMolfileWithAtomMap(entry.molfile3D);

  const topicMolecule = new TopicMolecule(molecule);
  topicMolecule.ensureMapNo();
  const hoses = topicMolecule.hoseCodes.map((hose) =>
    hose.map((idCode, sphere) => ({ idCode, sphere })),
  );
  entry.atoms = [];
  for (let i = 0; i < map.length; i++) {
    const sourceAtom = map[i];
    entry.atoms.push({ ...atoms[sourceAtom], hoses: hoses[i] });
  }
  entry.mf = molecule.getMolecularFormula().formula;
  entry.mw = molecule.getMolecularFormula().relativeWeight;
  entry.molfile2D = topicMolecule.toMolfileWithH();
  const { idCode, coordinates } = molecule.getIDCodeAndCoordinates();
  entry.idCode = idCode;
  entry.coordinates = coordinates;
  entry.ssIndex = getSSIndex(molecule);
  if (options?.date) entry.date = options.date;
  return entry;
}

function getSSIndex(molecule) {
  return Buffer.from(Uint32Array.from(molecule.getIndex()).buffer);
}
