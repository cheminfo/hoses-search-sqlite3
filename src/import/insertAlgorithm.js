import { getContactID } from './getContactID.js';
import { insertContact } from './insertContact.js';

export function insertAlgorithm(algorithm, contactMail, db) {
  let contactID = getContactID(contactMail, db);
  if (contactID === null) contactID = insertContact(contactMail, db);

  const checkIfExistsStmt = db.prepare(
    `SELECT * FROM algorithms WHERE 
    algorithms.name = '${algorithm.name}' AND 
    algorithms.version = '${algorithm.version}' AND 
    algorithms.contactID = ${contactID}`,
  );
  const similarRecords = checkIfExistsStmt.all();

  if (similarRecords.length === 0) {
    const insertStmt = db.prepare(`INSERT INTO 
    algorithms(name, version, description, contactID) 
    VALUES (?, ?, ?, ?)`);
    const newAlgorithmRecord = insertStmt.run(
      algorithm.name,
      algorithm.version,
      algorithm.description,
      contactID,
    );
    return newAlgorithmRecord.lastInsertRowid;
  } else {
    return similarRecords[0].algorithmID;
  }
}
