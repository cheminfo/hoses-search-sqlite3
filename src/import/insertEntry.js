/**
 * @param algorithmID
 * @param {InstanceType<import('better-sqlite3')>} db - the sqlite3 database
 * @param {object} entry - raw xyz data
 * @param {number} algorithmID - algorithm ID
 * @returns {number} - the value of the parameter
 */
export function insertEntry(entry, algorithmID, db) {
  const checkIfExistsStmt = db.prepare(`SELECT * FROM entries
    WHERE idCode='${entry.idCode}' AND algorithmID=${algorithmID}`);
  const similarRecords = checkIfExistsStmt.all();

  if (similarRecords.length === 0) {
    const insertStmt = db.prepare(`INSERT INTO 
        entries(mf, idCode, algorithmID, nbAtoms, coordinates, mw, comment, ssIndex, molfile2D, molfile3D, xyz, lastModificationDate) 
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`);
    const date = entry?.date ? Date.parse(entry.date) : Date.now();
    const newEntryRecord = insertStmt.run(
      entry.mf,
      entry.idCode,
      algorithmID,
      entry.nbAtoms,
      entry.coordinates,
      entry.mw,
      entry.comment,
      entry.ssIndex,
      entry.molfile2D,
      entry.molfile3D,
      entry.xyz,
      date,
    );
    return newEntryRecord.lastInsertRowid;
  } else {
    return similarRecords[0].entryID;
  }
}
