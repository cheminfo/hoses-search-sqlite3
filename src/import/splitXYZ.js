import { ensureString } from 'ensure-string';

export function splitXYZ(content) {
  content = ensureString(content);
  let entries = [];
  let entry;
  let lines = content.split('\n');
  for (let line of lines) {
    if (line.match(/^[0-9]+$/)) {
      if (entry !== undefined) entries.push(entry);
      entry = [];
    }
    if (!entry) {
      throw new Error('No number of atoms found in the first line');
    }
    entry.push(line);
  }

  return entries.filter((entry) => entry.length > 0);
}
