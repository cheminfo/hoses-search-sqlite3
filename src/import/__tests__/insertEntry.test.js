import { readFileSync } from 'node:fs';

import { test, expect } from 'vitest';

import { getTempDB } from '../../db/getDB.js';
import { getXYZEnhancedEntry } from '../getXYZEnhancedEntry.js';
import { insertAlgorithm } from '../insertAlgorithm.js';
import { insertEntry } from '../insertEntry.js';
import { splitXYZ } from '../splitXYZ.js';

test('insertEntry', async () => {
  const db = await getTempDB();
  expect(db.open).toBeTruthy();

  const xyzRawData = readFileSync(
    new URL('data/test.xyz', import.meta.url).pathname,
    'utf8',
  );

  const options = {
    contact: {
      email: 'test@test.ch',
    },
    xyz: {
      columns: [
        {
          column: 4,
          orbital: '2s',
          algorithm: {
            name: 'GW_charged',
            version: '1.0.0',
            description: 'GW_charged',
          },
        },
        {
          column: 5,
          orbital: '2s',
          algorithm: {
            name: 'dKS_charged',
            version: '1.0.0',
            description: 'dKS_charged',
          },
        },
        {
          column: 6,
          orbital: '2s',
          algorithm: {
            name: 'dKS_neutral',
            version: '1.0.0',
            description: 'dKS_neutral',
          },
        },
      ],
    },
  };

  const xyzEntries = splitXYZ(xyzRawData);
  const entry = await getXYZEnhancedEntry(xyzEntries[0], options);
  console.log(entry.atoms);

  for (let i = 0; i < entry.atoms.length; i++) {
    for (let p = 0; p < entry.atoms[i].properties.length; p++) {
      const property = entry.atoms[i].properties[p];
      if (!Number.isNaN(property.energy)) {
        const algorithmID = insertAlgorithm(
          property.algorithm,
          property.contact,
          db,
        );
        // Will import the contact if needed at the same time
        insertEntry(entry, algorithmID, db);
      }
    }
  }

  const stmt = db.prepare('SELECT * FROM entries');
  const insertedEntries = stmt.all();
  console.log(insertedEntries);
  for (let e = 0; e < insertedEntries.length; e++) {
    delete insertedEntries[e].lastModificationDate;
  }
  expect(insertedEntries).toMatchSnapshot();
});
