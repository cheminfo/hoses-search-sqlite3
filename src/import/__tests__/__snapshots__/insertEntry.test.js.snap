// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`insertEntry 1`] = `
[
  {
    "algorithmID": 1,
    "comment": "Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"",
    "coordinates": "#qeSE??}LAcq~\\Je\\AOcIR@EhAT\`D",
    "entryID": 1,
    "idCode": "fI@@",
    "mf": "H2O",
    "molfile2D": "Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
Actelion Java MolfileCreator 1.0

  3  2  0  0  0  0  0  0  0  0999 V2000
    6.9671    8.4598    6.9778 O   0  0  0  0  0  0  0  0  0  1  0  0
    7.1148    6.9671    6.9685 H   0  0  0  0  0  0  0  0  0  2  0  0
    8.3776    8.9702    6.9671 H   0  0  0  0  0  0  0  0  0  3  0  0
  2  1  1  0  0  0  0
  3  1  1  0  0  0  0
M  END
",
    "molfile3D": "Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
 OpenBabel10292409443D
Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
  3  2  0  0  0  0  0  0  0  0999 V2000
    4.5954    4.5000    4.5009 H   0  0  0  0  0  0  0  0  0  0  0  0
    4.5000    5.4641    4.5069 O   0  0  0  0  0  0  0  0  0  0  0  0
    5.4110    5.7938    4.5000 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  3  2  1  0  0  0  0
M  END
",
    "mw": 18.01488,
    "nbAtoms": 3,
    "ssIndex": {
      "data": [
        0,
        0,
        0,
        16,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "type": "Buffer",
    },
    "xyz": "3
Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
H 4.59540796 4.50000000 4.50085045
O 4.50000000 5.46412774 4.50692028
H 5.41100791 5.79382445 4.50000000",
  },
  {
    "algorithmID": 2,
    "comment": "Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"",
    "coordinates": "#qeSE??}LAcq~\\Je\\AOcIR@EhAT\`D",
    "entryID": 2,
    "idCode": "fI@@",
    "mf": "H2O",
    "molfile2D": "Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
Actelion Java MolfileCreator 1.0

  3  2  0  0  0  0  0  0  0  0999 V2000
    6.9671    8.4598    6.9778 O   0  0  0  0  0  0  0  0  0  1  0  0
    7.1148    6.9671    6.9685 H   0  0  0  0  0  0  0  0  0  2  0  0
    8.3776    8.9702    6.9671 H   0  0  0  0  0  0  0  0  0  3  0  0
  2  1  1  0  0  0  0
  3  1  1  0  0  0  0
M  END
",
    "molfile3D": "Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
 OpenBabel10292409443D
Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
  3  2  0  0  0  0  0  0  0  0999 V2000
    4.5954    4.5000    4.5009 H   0  0  0  0  0  0  0  0  0  0  0  0
    4.5000    5.4641    4.5069 O   0  0  0  0  0  0  0  0  0  0  0  0
    5.4110    5.7938    4.5000 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  3  2  1  0  0  0  0
M  END
",
    "mw": 18.01488,
    "nbAtoms": 3,
    "ssIndex": {
      "data": [
        0,
        0,
        0,
        16,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "type": "Buffer",
    },
    "xyz": "3
Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
H 4.59540796 4.50000000 4.50085045
O 4.50000000 5.46412774 4.50692028
H 5.41100791 5.79382445 4.50000000",
  },
  {
    "algorithmID": 3,
    "comment": "Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"",
    "coordinates": "#qeSE??}LAcq~\\Je\\AOcIR@EhAT\`D",
    "entryID": 3,
    "idCode": "fI@@",
    "mf": "H2O",
    "molfile2D": "Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
Actelion Java MolfileCreator 1.0

  3  2  0  0  0  0  0  0  0  0999 V2000
    6.9671    8.4598    6.9778 O   0  0  0  0  0  0  0  0  0  1  0  0
    7.1148    6.9671    6.9685 H   0  0  0  0  0  0  0  0  0  2  0  0
    8.3776    8.9702    6.9671 H   0  0  0  0  0  0  0  0  0  3  0  0
  2  1  1  0  0  0  0
  3  1  1  0  0  0  0
M  END
",
    "molfile3D": "Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
 OpenBabel10292409443D
Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
  3  2  0  0  0  0  0  0  0  0999 V2000
    4.5954    4.5000    4.5009 H   0  0  0  0  0  0  0  0  0  0  0  0
    4.5000    5.4641    4.5069 O   0  0  0  0  0  0  0  0  0  0  0  0
    5.4110    5.7938    4.5000 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  3  2  1  0  0  0  0
M  END
",
    "mw": 18.01488,
    "nbAtoms": 3,
    "ssIndex": {
      "data": [
        0,
        0,
        0,
        16,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "type": "Buffer",
    },
    "xyz": "3
Lattice="9.91100791 0.0 0.0 0.0 10.29382445 0.0 0.0 0.0 9.00692028" Properties=species:S:1:pos:R:3:GW_charged:R:1:dKS_charged:R:1:dKS_neutral:R:1 pbc="T T T"
H 4.59540796 4.50000000 4.50085045
O 4.50000000 5.46412774 4.50692028
H 5.41100791 5.79382445 4.50000000",
  },
]
`;
